### ISSUE TITLE
[Reproduction Issue] Possible batch-slot mismatch in runtime memory and configuration clarifications

### ISSUE THREAD

**Issue by JLUbaijh2323** [InternRobotics/EventVLA#7]
</p><hr><h1>Possible runtime memory batching issue and configuration clarification for RoboTwin-MeM reproduction</h1><p>Hi EventVLA authors,</p><p>Thank you very much for releasing EventVLA, RoboTwin-MeM benchmark, checkpoints, and training/evaluation code.</p><p>We are currently reproducing EventVLA on RoboTwin-MeM and have been carefully checking the training pipeline. During reproduction, we noticed several behaviors that may affect runtime keyframe memory propagation and the final performance.</p><p>We are not sure whether these are expected behaviors or differences between the public release and the internal training setup, so we would like to ask for clarification.</p><p>Our reproduction is based on:</p><pre><code>Repository:
https://github.com/InternRobotics/EventVLA

Commit:
4b5b26030abddf83bc60e1a6b067de8f521fd0ec
</code></pre><hr><h2>1. Possible mismatch between runtime memory and batch slot ordering</h2><p>From our understanding of EventVLA's runtime memory mechanism, predicted keyframes are maintained according to batch slots.</p><p>The runtime memory contains states such as:</p><pre><code class="language-python">_runtime_keyframe_image_bank[slot_idx]
_runtime_pending_keyframe_writes[slot_idx]
_runtime_slot_episode_ids[slot_idx]
</code></pre><p>Therefore, we assume that each batch slot should maintain a causal trajectory:</p><pre><code>Batch 0:
slot 0 -&gt; episode A, t=0
slot 1 -&gt; episode B, t=0
slot 2 -&gt; episode C, t=0
...

Batch 1:
slot 0 -&gt; episode A, t=50
slot 1 -&gt; episode B, t=50
slot 2 -&gt; episode C, t=50
...
</code></pre><p>so that the keyframe predicted at an earlier timestep can be consumed by later timesteps of the same episode.</p><p>However, when using the current public <code inline="">SequentialEpisodeBatchSampler</code>, we observed that samples seem to be flattened first and then split into batches.</p><p>For example:</p><pre><code>Batch 0:
slot 0 -&gt; episode A, t=0
slot 1 -&gt; episode A, t=50
slot 2 -&gt; episode A, t=100
slot 3 -&gt; episode A, t=150

Batch 1:
slot 0 -&gt; episode A, t=200
slot 1 -&gt; episode A, t=250
slot 2 -&gt; episode A, t=300
slot 3 -&gt; episode A, t=350
</code></pre><p>In this case, the temporal sequence of one memory slot becomes:</p><pre><code>slot 0:
t=0 -&gt; t=200 -&gt; t=400 ...
</code></pre><p>instead of:</p><pre><code>t=0 -&gt; t=50 -&gt; t=100 ...
</code></pre><p>This may affect:</p><ol><li><p>predicted keyframe propagation;</p></li><li><p>exact keyframe fetching;</p></li><li><p>pending memory write consumption;</p></li><li><p>runtime memory bank population after teacher forcing decreases.</p></li></ol><p>Especially with:</p><pre><code class="language-yaml">keyframe_train_memory_source: teacher_to_predict
keyframe_eval_memory_source: predict
</code></pre><p>a keyframe predicted at timestep <code inline="">t</code> may not be consumed by the next causal timestep because the next timestep may already be located in another batch slot.</p><hr><h2>2. Experiment: persistent episode-slot sampler</h2><p>To verify this hypothesis, we implemented an optional sampler modification.</p><p>Instead of assigning arbitrary samples to batch slots, we preserve episode-slot correspondence:</p><pre><code>Batch 0:

slot 0 -&gt; episode A, t=0
slot 1 -&gt; episode B, t=0
slot 2 -&gt; episode C, t=0

Batch 1:

slot 0 -&gt; episode A, t=50
slot 1 -&gt; episode B, t=50
slot 2 -&gt; episode C, t=50
</code></pre><p>This allows runtime memory to evolve causally.</p><p>We compared three settings:</p><ol><li><p>Original public reproduction.</p></li><li><p>Oracle GT memory (always using ground-truth keyframes).</p></li><li><p>Modified persistent episode-slot sampler.</p></li></ol><hr><h3><img width="696" height="516" alt="Image" src="https://github.com/user-attachments/assets/c152a887-ad53-4fd0-bb43-dcd2cecdd0ee" />

Figure 1: Original reproduction vs GT memory oracle</h3><p>Green:<br>Original reproduction with predicted runtime memory.</p><p>Yellow:<br>Always using GT keyframe memory.</p><p>Observation:</p><p>The original reproduction shows a significant reduction of runtime memory keyframe count after switching from teacher memory to predicted memory, while the GT oracle remains stable.</p><hr><h3>

<img width="898" height="476" alt="Image" src="https://github.com/user-attachments/assets/bbba2f48-ca9d-46c6-896f-4a03b66ac6c2" />

Figure 2: Modified persistent episode-slot sampler</h3><p>Blue:<br>Persistent episode-slot sampler.</p><p>Observation:</p><p>After preserving episode-slot causality, runtime memory remains much more stable and behaves closer to the GT setting.</p><hr><p>Based on these observations, we would like to ask:</p><ol><li><p>Was the original EventVLA training performed with a sampler that preserves episode-to-batch-slot correspondence?</p></li><li><p>Does the released <code inline="">SequentialEpisodeBatchSampler</code> represent the exact training sampler used for the paper?</p></li><li><p>Should runtime memory slots always correspond to the same episode across consecutive batches when using <code inline="">teacher_to_predict</code>?</p></li><li><p>Is there any additional mechanism in the internal code that handles this memory-slot propagation?</p></li></ol><hr><h1>3. Configuration differences between paper, YAML, and released scripts</h1><p>During reproduction, we also noticed some differences between the paper description, default YAML configuration, and launcher scripts.</p><p>For example:</p>
Item | Paper / checkpoint description | Public YAML
-- | -- | --
Training steps | 80k | 80k
Released checkpoint | steps_100000 | 100k
Memory size | 5 | 4
Batch size | 4 | 8
Keyframe NMS | unclear | 20
Keyframe cooldown | unclear | 20

<p>We are unsure whether these correspond to different experimental versions.</p><p>Could you please clarify the exact configuration used for the reported RoboTwin-MeM results?</p><p>Especially:</p><ul><li><p>global batch size;</p></li><li><p>number of GPUs;</p></li><li><p>training steps;</p></li><li><p>maximum memory keyframes;</p></li><li><p>keyframe loss weight;</p></li><li><p>teacher-to-predict schedule;</p></li><li><p>sampling interval.</p></li></ul><hr><h1>4. Dataset mixture clarification</h1><p>We noticed that the default:</p><pre><code class="language-yaml">data_mix: robotwin_mem
</code></pre><p>does not exactly match the eight tasks described in the README.</p><p>The README lists:</p><pre><code>cover_blocks_hard
find_seal_and_seal_stamp
pick_objects_in_order
pick_the_unhidden_block
press_button_keyframe
put_back_block_hard
rearrange_blocks_hard
reproduce_route
</code></pre><p>However, the current public <code inline="">robotwin_mem</code> mixture contains:</p><pre><code>cover_blocks_hard
put_back_block_hard
rearrange_blocks_hard
observe_and_pickup_hard
find_seal_and_seal_stamp
observe_and_pickup_object
reproduct_route
press_button_keyframe
</code></pre><p>We noticed:</p><ul><li><p><code inline="">pick_objects_in_order</code> is replaced by <code inline="">observe_and_pickup_hard</code>;</p></li><li><p><code inline="">pick_the_unhidden_block</code> is replaced by <code inline="">observe_and_pickup_object</code>;</p></li><li><p><code inline="">reproduct_route</code> may be a typo of <code inline="">reproduce_route</code>.</p></li></ul><p>The launcher seems to use <code inline="">robotwin_mem8</code>, but this mixture contains author-specific absolute paths.</p><p>Could you please clarify:</p><ol><li><p>Which dataset mixture was used for the released checkpoint?</p></li><li><p>Is the README eight-task list the correct benchmark training set?</p></li><li><p>Should <code inline="">robotwin_mem8</code> be considered the official training mixture?</p></li></ol><hr><h1>5. QwenOFT naming clarification</h1><p>We also noticed a possible naming inconsistency.</p><p>The paper describes EventVLA as based on QwenOFT, while the public framework entry uses:</p><pre><code class="language-yaml">framework:
  name: EventVLA
</code></pre><p>At the same time, some evaluation scripts still contain names such as:</p><pre><code>weights_8tasks_pure_image_keyframe_memory_teacher_qwenoft.sh
</code></pre><p>Could you please clarify:</p><ol><li><p>Is QwenOFT the original VLA backbone internally used by EventVLA?</p></li><li><p>Is <code inline="">framework.name=EventVLA</code> the correct public entry point?</p></li><li><p>Is the <code inline="">qwenoft</code> naming only historical/legacy naming?</p></li></ol><hr><p>Thank you again for releasing this work.</p><p>We appreciate any clarification about the intended training pipeline.<br>If needed, we can also provide:</p><ul><li><p>modified sampler implementation;</p></li><li><p>detailed batch-slot / episode-slot logs;</p></li><li><p>runtime memory statistics;</p></li><li><p>additional reproduction curves.</p></li></ul><h2>Thank you very much!</h2>

**Comment by zhangzhengtu** [comment]
Hi, thank you very much for the detailed reproduction report and analysis.

May I ask whether you have already trained EventVLA on the RoboTwin-MeM dataset with the modified persistent episode-slot sampler? If so, did you obtain reasonably good success rates compared with the released checkpoint or the reported paper results?

It would be very helpful if you could share the per-task success rates, or at least the overall average success rate, under this setting. This can help us better understand whether preserving episode-to-batch-slot causality is indeed critical for reproducing the reported RoboTwin-MeM performance.

Thanks again for your careful investigation and for sharing these findings.

**Comment by JLUbaijh2323** [comment]
Thanks for following up. Yes, I have trained EventVLA on RoboTwin-MeM with the modified persistent episode-slot sampler described above.

Following the released 100k training setup, I evaluated the reproduced model on all 8 RoboTwin-MeM tasks and obtained an average success rate of 76.3%:

Hidden: 76% | Order: 46% | Cover: 100% | Seal: 72% | PutBack: 100% | Press: 54% | Rearrange: 78% | Route: 84%

The experiment was run on 16 PPUs with a per-device batch size of 4.

For comparison, before introducing the persistent episode-slot modification, I remember the overall success rate being significantly below 60%, although unfortunately I did not keep the exact number from that run.

So far, the results suggest that preserving episode-to-batch-slot causality has a substantial impact on the reproduction performance. This also seems consistent with the runtime-memory statistics shown above, where the modified sampler maintains a much more stable predicted keyframe memory. However, I would not conclude from this experiment alone that the sampler is the only source of the reproduction gap.

**Comment by zhangzhengtu** [comment]
Thank you very much for the detailed reproduction and experimental results. The 76.3% success rate, together with the runtime memory statistics, further suggests that preserving episode-to-batch-slot causality is indeed important.

We will adopt your sampler modification and also fix several small issues and configuration inconsistencies in the current public release. After verification, we will release a complete updated version of the code.

Thanks again for your careful investigation and contribution!
